idaes/apps/caprese/estimator.py

```diff
@@ -0,0 +1,597 @@
+        for var, val in model_disturbance_weights:  
+            #Check whether the given variable is classified under diffvar
+            if id(var) not in diff_id_list:
+                raise RuntimeError(var.name, 
+                                   " is not a differential variable.")
+
+            if givenform == "weight":
+                self.diffvar_map_moddis[var].weight = val
+            elif givenform == "variance":
+                self.diffvar_map_moddis[var].weight = 1./val
+
+        mea_id_list = [id(var[0]) for var in self.measurement_vars]
+        for var, val in measurement_noise_weights:
+            #Check whether the given variable is declared as a measurement before
+            if id(var) not in mea_id_list:
+                raise RuntimeError(var.name, 
+                                   " is not declared as a measurement.")
+
+            if givenform == "weight":
+                self.meavar_map_meaerr[var].weight = val
+            elif givenform == "variance":
+                self.meavar_map_meaerr[var].weight = 1./val
+
+        moddis_block = self.MODELDISTURBANCE_BLOCK
+        moddis_list = [moddis_block[ind].var for ind in self.DIFFERENTIAL_SET]
+
+        meaerr_block = self.MEASUREMENTERROR_BLOCK
+        meaerr_list = [meaerr_block[ind].var for ind in self.MEASUREMENT_SET]
+
+        wQw = sum(
+            moddis.weight * moddis[sampt]**2
+            for moddis in moddis_list
+            for sampt in self.sample_points
+            if sampt != self.time.first() #Skip model disturbnace at time.first()
+            )
+
+        vRv = sum(
+            meaerr.weight * meaerr[sampt]**2
+            for meaerr in meaerr_list
+            for sampt in self.sample_points #Here should include the time.first()
+            )
+
+        self.noise_minimize_objective = Objective(expr = (wQw) + (vRv))  
+
+    def check_var_con_dof(self, skip_dof_check = False):
+        '''This function does the final check for fixed and unfixed variables
+        as well as the deactivated differential equations. Then check the degree of freedom.
+
+        Parameters
+        ----------
+        skip_dof_check: SKip to check the degree of freedom if it's True. 
+                        Default is False.
+        '''
+
+        self.vectors.input[...].fix()
+        self.vectors.differential[...].unfix()
+        self.vectors.modeldisturbance[:,0].fix(0.0)
+
+        for diff_eq in self.con_category_dict[CC.DIFFERENTIAL]:
+            diff_eq.deactivate()
+
+        n_diffvars = len(self.DIFFERENTIAL_SET.ordered_data())
+        n_steps_in_horizon = len(self.sample_points) - 1
+        correct_dof = n_diffvars * (n_steps_in_horizon + 1)
+
+        if not skip_dof_check:
+            dof = degrees_of_freedom(self)
+            assert dof == correct_dof  
+
+    def load_inputs_for_MHE(self, inputs):
+        last_sampt = self.sample_points[-1]
+        secondlast_sampt = self.sample_points[-2]
```

Comment:
Why do we load inputs between the last and second to last samples?

Reply:
It's because the latest `actualmeasurement` is generated from the plant with the latest control input. Therefore, that latest and known control input should be sent to the estimator as well.

Reply:
Okay I see, that makes perfect sense.